### PR TITLE
bpo-33330: PyImport_Cleanup check for exc leaks

### PR DESCRIPTION
**Pull request by vstinner**
* Add "assert(!PyErr_Occurred());" assertions to PyImport_Cleanup()
  and _PyGC_CollectNoFail() to make sure that these functions don't
  leak exceptions.
* Replace also PyImport_GetModuleDict() call with interp->modules to
  prevent a fatal error if modules is already NULL.

https://bugs.python.org/issue33330

### CODE REVIEW

**Comment by vstinner** [comment]
The commit d393c1b227f22fb9af66040b2b367c99a4d1fa9a replaced interp->modules with PyImport_GetModuleDict(), but this function fails with a fatal error if modules is NULL. I'm not sure that it was a deliberate change. Moreover, this commit has been partially reverted (interp->modules is back!).

**Comment by vstinner** [comment]
PyImport_Cleanup() is part of the public API, but it's documented as "Empty the module table. For internal use only." :-( Why do we have public API for internal use only? Likely a mistake from the past.

File: Python/import.c
Comment (serhiy-storchaka) [inline Python/import.c:399]: I'm wondering if it is worth to call `PyErr_WriteUnraisable()` here. This will expose some information about the exception raised outside of this function.

File: Modules/gcmodule.c
Comment (serhiy-storchaka) [inline Modules/gcmodule.c:1598]: Yesterday I wrote a patch that improves errors handling in the garbage collector. bpo-33612 is the result of testing it. And I have added similar asserts in `_PyGC_CollectNoFail()` (but the patch is not completed yet, maybe I'll move them).

At least move the second assert inside the "else" block.

File: Python/import.c
Comment (serhiy-storchaka) [inline Python/import.c:484]: Neither `PyDict_Next()` nor any other calls before can raise an exception.

File: Python/import.c
Comment (serhiy-storchaka) [inline Python/import.c:500]: `PyObject_GetIter()`, `PyIter_Next()` and `PyObject_GetItem()` can raise and exception if corresponding slots are implemented incorrectly in extensions. `assert` looks good here.

File: Python/import.c
Comment (serhiy-storchaka) [inline Python/import.c:511]: This never fails.

File: Python/import.c
Comment (serhiy-storchaka) [inline Python/import.c:534]: This never fails.

File: Python/import.c
Comment (serhiy-storchaka) [inline Python/import.c:555]: This never fails.

File: Python/import.c
Comment (serhiy-storchaka) [inline Python/import.c:572]: This never fails.

**Comment by serhiy-storchaka** [comment]
I would left only few asserts after calls that can raise exceptions if the third-party code has bugs. And one at the end of `PyImport_Cleanup()` just for sure. For other asserts we can guarantee the absence of exceptions.

File: Python/import.c
Comment (vstinner) [reply Python/import.c:484]: This is the theory. In practice, something called before might leak an exception. It's to catch bugs "just in case". The assertion is cheap. Debugging an "exception leak" can be very painful.

**Comment by vstinner** [comment]
The last 2 years, I added many "assert(!PyErr_Occurred());" in ceval.c and object.c, and it helped me a lot to detect some very old bugs. The rationale here is the same.

File: Python/import.c
Comment (vstinner) [reply Python/import.c:399]: The contract of this function is that the caller must not hold an exception. If the assertion fails, I would prefer to fix the caller rather than working around the bug using PyErr_WriteUnraisable().

The point is not only to catch bugs in the current code, but also ease debug when the callers code will be modified. Python shutdown process is very fragile and error prone.

File: Python/import.c
Comment (serhiy-storchaka) [reply Python/import.c:484]: There is no code between this line and previous error check that can raise an exception and don't silence it. I have checked this.

File: Python/import.c
Comment (ericsnowcurrently) [inline Python/import.c:407]: I'd rather avoid this, if possible.  Under what conditions does the use of PyImport_GetModuleDict() cause a fatal error?

Comment (vstinner) [reply Python/import.c:407]: If you call PyImport_Cleanup() twice. In Python 3.6, calling PyImport_Cleanup() a second time does nothing: that's the purpose of the test on the next line.

PyImport_Cleanup() is part of the public API, even if it's documented as "please don't call this function" ;-) I would prefer to restore Python 3.6 behaviour.

Comment (ericsnowcurrently) [reply Python/import.c:407]: Ah, thanks for clarifying.  LGTM then.  I still plan on landing #3606 at some point, but at that point I'll have to deal with recent ``interp->modules`` usage anyway. :)

**Comment by serhiy-storchaka** [comment]
See also #7126.

**Review by serhiy-storchaka — COMMENTED** [review]
LGTM, except that GC is covered by a separate PR.

**Review by serhiy-storchaka — APPROVED** [review]

**Comment by vstinner** [comment]
> LGTM, except that GC is covered by a separate PR.

I will wait until #7126 is merged, and then rebase this PR on top of it.

**Comment by serhiy-storchaka** [comment]
Why not rebase and merge this PR?

**Comment by vstinner** [comment]
It's just that I wrote it 2 months ago, and I lost interest in this minor enhancement in the meanwhile. I'm trying to reduce my number of open PRs. Feel free to reuse my patch to create a new PR if you want.

**Pull request closed without merging**
